### ISSUE TITLE
Upgrade to NAN 2 for Node 4 Compatibility

### ISSUE THREAD

**Issue by mrapczynski** [oracle/node-oracledb#115]
Working with this driver package using "real" node, and also io.js. To be clear - it works in **both** flavors of Node.

However, the build using the IO.js flavored npm yields some scary looking deprecation warnings that maybe you would like to know about for the future.

```shell
[root@dockern1 test]# npm install oracledb
|
> oracledb@0.6.0 install /opt/test/node_modules/oracledb
> node-gyp rebuild

gyp WARN EACCES user "root" does not have permission to access the dev dir "/root/.node-gyp/2.3.0"
gyp WARN EACCES attempting to reinstall using temporary dev dir "/opt/test/node_modules/oracledb/.node-gyp"
make: Entering directory '/opt/test/node_modules/oracledb/build'
  CXX(target) Release/obj.target/oracledb/src/njs/src/njsOracle.o
../src/njs/src/njsOracle.cpp: In static member function ‘static void Oracledb::Async_AfterGetConnection(uv_work_t*)’:
../src/njs/src/njsOracle.cpp:602:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsOracle.cpp: In static member function ‘static void Oracledb::Async_AfterCreatePool(uv_work_t*)’:
../src/njs/src/njsOracle.cpp:753:29: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException (tc);
                             ^
  CXX(target) Release/obj.target/oracledb/src/njs/src/njsPool.o
../src/njs/src/njsPool.cpp: In static member function ‘static void Pool::Async_AfterGetConnection(uv_work_t*)’:
../src/njs/src/njsPool.cpp:491:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsPool.cpp: In static member function ‘static void Pool::Async_AfterTerminate(uv_work_t*)’:
../src/njs/src/njsPool.cpp:594:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
  CXX(target) Release/obj.target/oracledb/src/njs/src/njsConnection.o
../src/njs/src/njsConnection.cpp: In static member function ‘static void Connection::Async_AfterExecute(uv_work_t*)’:
../src/njs/src/njsConnection.cpp:1096:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsConnection.cpp: In static member function ‘static void Connection::Async_AfterRelease(uv_work_t*)’:
../src/njs/src/njsConnection.cpp:1537:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsConnection.cpp: In static member function ‘static void Connection::Async_AfterCommit(uv_work_t*)’:
../src/njs/src/njsConnection.cpp:1630:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsConnection.cpp: In static member function ‘static void Connection::Async_AfterRollback(uv_work_t*)’:
../src/njs/src/njsConnection.cpp:1722:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsConnection.cpp: In static member function ‘static void Connection::Async_AfterBreak(uv_work_t*)’:
../src/njs/src/njsConnection.cpp:1816:28: warning: ‘void node::FatalException(const v8::TryCatch&)’ is deprecated (declared at /opt/test/node_modules/oracledb/.node-gyp/2.3.0/src/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
  CXX(target) Release/obj.target/oracledb/src/njs/src/njsMessages.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiEnv.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiEnvImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiException.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiExceptionImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiConnImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiDateTimeArrayImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiPoolImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiStmtImpl.o
  CXX(target) Release/obj.target/oracledb/src/dpi/src/dpiUtils.o
  SOLINK_MODULE(target) Release/obj.target/oracledb.node
  COPY Release/oracledb.node
make: Leaving directory '/opt/test/node_modules/oracledb/build'
oracledb@0.6.0 node_modules/oracledb
└── nan@1.8.4
```

**Comment by cjbj** [comment]
@mrapczynski  Thanks for the report.  What Node.js/io.js version, OS distribution and compiler do you have?

**Comment by mrapczynski** [comment]
@cjbj You're welcome. See below.

```shell
[root@dockern1 ~]# iojs -v
v2.3.0

[root@dockern1 ~]# npm -v
2.11.1

[root@dockern1 ~]# gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.9.2/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.9.2-20150212/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.9.2-20150212/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.9.2 20150212 (Red Hat 4.9.2-6) (GCC)

[root@dockern1 ~]# cat /etc/redhat-release
Fedora release 21 (Twenty One)
```

**Comment by cjbj** [comment]
@mrapczynski thanks.

**Comment by Bigous** [comment]
Just to add some information... it started with IO.js 2.0 - [witch is not officially supported by NAN yet](https://github.com/nodejs/nan/blob/master/CHANGELOG.md).
It's normally handled by NaN and, probably we will have to do minor efforts when it becomes true.

**Comment by cjbj** [comment]
@Bigous thanks for keeping track.

**Comment by Bigous** [comment]
Ok, [now NaN does provide support for more recent IO.js](https://github.com/nodejs/nan/blob/master/CHANGELOG.md), but they jumped from 1.X to 3.X.
The problem with that is the backward compatibility break. They removed some macros that we use, like NanReturn.
I'll make a new pull request to upgrade the usage of NaN to 2.x.x soon.

**Comment by cjbj** [comment]
@Bigous that's great it's much appreciated. I did notice NAN 2 broke our build.  Our next drop will still use NAN 1.

**Comment by bpjones** [comment]
I tried to install using node v4 today and wasn't successful...
```bash
txhous10apl28:test bjones5$ npm install oracledb --save
\
> oracledb@1.1.0 install /Users/bjones5/Documents/node/test/node_modules/oracledb
> node-gyp rebuild

  CXX(target) Release/obj.target/oracledb/src/njs/src/njsOracle.o
In file included from ../src/njs/src/njsOracle.cpp:53:
In file included from ../src/njs/src/njsOracle.h:56:
../node_modules/nan/nan.h:324:27: error: redefinition of 'NanEnsureHandleOrPersistent'
  NAN_INLINE v8::Local<T> NanEnsureHandleOrPersistent(const v8::Local<T> &val) {
                          ^
../node_modules/nan/nan.h:319:17: note: previous definition is here
  v8::Handle<T> NanEnsureHandleOrPersistent(const v8::Handle<T> &val) {
                ^
../node_modules/nan/nan.h:344:27: error: redefinition of 'NanEnsureLocal'
  NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Handle<T> &val) {
                          ^
../node_modules/nan/nan.h:334:27: note: previous definition is here
  NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Local<T> &val) {
                          ^
../node_modules/nan/nan.h:757:13: error: no member named 'smalloc' in namespace 'node'
    , node::smalloc::FreeCallback callback
      ~~~~~~^
../node_modules/nan/nan.h:768:12: error: no matching function for call to 'New'
    return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
           ^~~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/node_buffer.h:31:40: note: candidate function not viable: no known
      conversion from 'uint32_t' (aka 'unsigned int') to 'enum encoding' for 3rd argument
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node_buffer.h:43:40: note: candidate function not viable: 2nd argument
      ('const char *') would lose const qualifier
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node_buffer.h:28:40: note: candidate function not viable: requires 2
      arguments, but 3 were provided
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate, size_t length);
                                       ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node_buffer.h:36:40: note: candidate function not viable: requires 5
      arguments, but 3 were provided
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
In file included from ../src/njs/src/njsOracle.cpp:53:
In file included from ../src/njs/src/njsOracle.h:56:
../node_modules/nan/nan.h:772:12: error: no viable conversion from 'v8::MaybeLocal<v8::Object>' to
      'v8::Local<v8::Object>'
    return node::Buffer::New(v8::Isolate::GetCurrent(), size);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:210:7: note: candidate constructor (the implicit copy constructor) not
      viable: no known conversion from 'v8::MaybeLocal<v8::Object>' to 'const v8::Local<v8::Object> &' for 1st argument
class Local {
      ^
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:210:7: note: candidate constructor (the implicit move constructor) not
      viable: no known conversion from 'v8::MaybeLocal<v8::Object>' to 'v8::Local<v8::Object> &&' for 1st argument
class Local {
      ^
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:214:13: note: candidate template ignored: could not match 'Local'
      against 'MaybeLocal'
  V8_INLINE Local(Local<S> that)
            ^
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:326:13: note: candidate template ignored: could not match 'S *' against
      'v8::MaybeLocal<v8::Object>'
  V8_INLINE Local(S* that)
            ^
In file included from ../src/njs/src/njsOracle.cpp:53:
In file included from ../src/njs/src/njsOracle.h:56:
../node_modules/nan/nan.h:779:26: error: no member named 'Use' in namespace 'node::Buffer'
    return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
           ~~~~~~~~~~~~~~^
../src/njs/src/njsOracle.cpp:761:11: warning: 'FatalException' is deprecated: Use FatalException(isolate, ...)
      [-Wdeprecated-declarations]
    node::FatalException(tc);
          ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node.h:283:29: note: 'FatalException' has been explicitly marked deprecated
      here
                inline void FatalException(const v8::TryCatch& try_catch) {
                            ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node.h:66:42: note: expanded from macro 'NODE_DEPRECATED'
    __attribute__((deprecated(message))) declarator
                                         ^
../src/njs/src/njsOracle.cpp:915:11: warning: 'FatalException' is deprecated: Use FatalException(isolate, ...)
      [-Wdeprecated-declarations]
    node::FatalException (tc);
          ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node.h:283:29: note: 'FatalException' has been explicitly marked deprecated
      here
                inline void FatalException(const v8::TryCatch& try_catch) {
                            ^
/Users/bjones5/.node-gyp/4.0.0/include/node/node.h:66:42: note: expanded from macro 'NODE_DEPRECATED'
    __attribute__((deprecated(message))) declarator
                                         ^
In file included from ../src/njs/src/njsOracle.cpp:51:
In file included from /Users/bjones5/.node-gyp/4.0.0/include/node/node.h:42:
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:221:5: error: assigning to 'v8::Primitive *volatile' from incompatible
      type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:501:12: note: in instantiation of function template specialization
      'v8::Local<v8::Primitive>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(v8::Undefined(v8::Isolate::GetCurrent())));
           ^
../node_modules/nan/nan.h:483:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(Nan::imp::NanEnsureLocal(val))
                             ^
In file included from ../src/njs/src/njsOracle.cpp:51:
In file included from /Users/bjones5/.node-gyp/4.0.0/include/node/node.h:42:
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:221:5: error: assigning to 'v8::Boolean *volatile' from incompatible
      type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:511:12: note: in instantiation of function template specialization
      'v8::Local<v8::Boolean>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(v8::True(v8::Isolate::GetCurrent())));
           ^
../node_modules/nan/nan.h:483:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(Nan::imp::NanEnsureLocal(val))
                             ^
In file included from ../src/njs/src/njsOracle.cpp:51:
In file included from /Users/bjones5/.node-gyp/4.0.0/include/node/node.h:42:
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:221:5: error: assigning to 'v8::Function *volatile' from incompatible
      type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:1645:12: note: in instantiation of function template specialization
      'v8::Local<v8::Function>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(handle)->Get(kCallbackIndex)
           ^
../node_modules/nan/nan.h:483:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(Nan::imp::NanEnsureLocal(val))
                             ^
In file included from ../src/njs/src/njsOracle.cpp:51:
In file included from /Users/bjones5/.node-gyp/4.0.0/include/node/node.h:42:
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:221:5: error: assigning to 'v8::Object *volatile' from incompatible
      type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/bjones5/.node-gyp/4.0.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:1776:12: note: in instantiation of function template specialization
      'v8::Local<v8::Object>::Local<v8::Value>' requested here
    return NanEscapeScope(
           ^
../node_modules/nan/nan.h:483:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(Nan::imp::NanEnsureLocal(val))
                             ^
2 warnings and 10 errors generated.
make: *** [Release/obj.target/oracledb/src/njs/src/njsOracle.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.5.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/bjones5/Documents/node/test/node_modules/oracledb
gyp ERR! node -v v4.0.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok
npm ERR! Darwin 14.5.0
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "oracledb" "--save"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE

npm ERR! oracledb@1.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the oracledb@1.1.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the oracledb package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls oracledb
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/bjones5/Documents/node/test/npm-debug.log
```

**Comment by cjbj** [comment]
@bpjones thanks for confirming the behavior with v4.

**Comment by ijaooo** [comment]
It compiles on Windows, but crashes node v4. On Linux, it doesn't even compile (same errors as earlier reports).

**Comment by Bigous** [comment]
I've ported to nan 2.x , but it crashes when using blobs specificaly 4.1.1 in the mocha.

I'm on vacation right now - but as soon as I get back, I'll go back to this issue

**Comment by binlaniua** [comment]
i have the same issue

**Comment by anupamasajja** [comment]
I have the same issue installing oracledb in Linux machine. I am using v4.0.0 node version

This is the error message that I am getting

npm install oracledb
npm WARN package.json myproject@1.0.0 No repository field.
npm WARN package.json myproject@1.0.0 No README data
-
> oracledb@1.1.0 install /home/dc-user/myproject/node_modules/oracledb
> node-gyp rebuild
 
make: Entering directory `/home/dc-user/myproject/node_modules/oracledb/build'
  CXX(target) Release/obj.target/oracledb/src/njs/src/njsOracle.o
In file included from ../src/njs/src/njsOracle.h:56:0,
                 from ../src/njs/src/njsOracle.cpp:53:
../node_modules/nan/nan.h:324:27: error: redefinition of âtemplate<class T> v8::
Local<T> Nan::imp::NanEnsureHandleOrPersistent(const v8::Local<T>&)â
   NAN_INLINE v8::Local<T> NanEnsureHandleOrPersistent(const v8::Local<T> &val)
{
                           ^
../node_modules/nan/nan.h:319:17: error: âtemplate<class T> v8::Handle<T> Nan::i
mp::NanEnsureHandleOrPersistent(v8::Handle<T>&)â previously declared here
   v8::Handle<T> NanEnsureHandleOrPersistent(const v8::Handle<T> &val) {
                 ^
../node_modules/nan/nan.h:344:27: error: redefinition of âtemplate<class T> v8::
Local<T> Nan::imp::NanEnsureLocal(v8::Handle<T>&)â
   NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Handle<T> &val) {
                           ^
../node_modules/nan/nan.h:334:27: error: âtemplate<class T> v8::Local<T> Nan::im
p::NanEnsureLocal(const v8::Local<T>&)â previously declared here
   NAN_INLINE v8::Local<T> NanEnsureLocal(const v8::Local<T> &val) {
                           ^
../node_modules/nan/nan.h:757:13: error: ânode::smallocâ has not been declared
     , node::smalloc::FreeCallback callback
             ^
../node_modules/nan/nan.h:757:35: error: expected â,â or â...â before âcallbackâ
     , node::smalloc::FreeCallback callback
                                   ^
../node_modules/nan/nan.h: In function âv8::Local<v8::Object> NanNewBufferHandle
(char*, size_t, int)â:
../node_modules/nan/nan.h:761:50: error: âcallbackâ was not declared in this sco
pe
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                  ^
../node_modules/nan/nan.h:761:60: error: âhintâ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                            ^
../node_modules/nan/nan.h: In function âv8::Local<v8::Object> NanNewBufferHandle
(const char*, uint32_t)â:
../node_modules/nan/nan.h:768:67: error: call of overloaded âNew(v8::Isolate*, c
onst char*&, uint32_t&)â is ambiguous
     return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
                                                                   ^
../node_modules/nan/nan.h:768:67: note: candidates are:
In file included from ../node_modules/nan/nan.h:25:0,
                 from ../src/njs/src/njsOracle.h:56,
                 from ../src/njs/src/njsOracle.cpp:53:
/home/dc-user/.node-gyp/4.0.0/include/node/node_buffer.h:31:40: note: v8::MaybeL
ocal<v8::Object> node::Buffer::New(v8::Isolate*, v8::Local<v8::String>, node::en
coding) <near match>
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/home/dc-user/.node-gyp/4.0.0/include/node/node_buffer.h:31:40: note:   no known
conversion for argument 3 from âuint32_t {aka unsigned int}â to ânode::encoding
â
/home/dc-user/.node-gyp/4.0.0/include/node/node_buffer.h:43:40: note: v8::MaybeL
ocal<v8::Object> node::Buffer::New(v8::Isolate*, char*, size_t) <near match>
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/home/dc-user/.node-gyp/4.0.0/include/node/node_buffer.h:43:40: note:   no known
conversion for argument 2 from âconst char*â to âchar*â
In file included from ../src/njs/src/njsOracle.h:56:0,
                 from ../src/njs/src/njsOracle.cpp:53:
../node_modules/nan/nan.h: In function âv8::Local<v8::Object> NanNewBufferHandle
(uint32_t)â:
../node_modules/nan/nan.h:772:61: error: could not convert ânode::Buffer::New(v8
::Isolate::GetCurrent(), ((size_t)size))â from âv8::MaybeLocal<v8::Object>â to â
v8::Local<v8::Object>â
     return node::Buffer::New(v8::Isolate::GetCurrent(), size);
                                                             ^
../node_modules/nan/nan.h: In function âv8::Local<v8::Object> NanBufferUse(char*
, uint32_t)â:
../node_modules/nan/nan.h:779:12: error: âUseâ is not a member of ânode::Bufferâ
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
            ^
../src/njs/src/njsOracle.cpp: In static member function âstatic void Oracledb::A
sync_AfterGetConnection(uv_work_t*)â:
../src/njs/src/njsOracle.cpp:761:28: warning: âvoid node::FatalException(const v
8::TryCatch&)â is deprecated (declared at /home/dc-user/.node-gyp/4.0.0/include/
node/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException(tc);
                            ^
../src/njs/src/njsOracle.cpp: In static member function âstatic void Oracledb::A
sync_AfterCreatePool(uv_work_t*)â:
../src/njs/src/njsOracle.cpp:915:29: warning: âvoid node::FatalException(const v
8::TryCatch&)â is deprecated (declared at /home/dc-user/.node-gyp/4.0.0/include/
node/node.h:282): Use FatalException(isolate, ...) [-Wdeprecated-declarations]
     node::FatalException (tc);
                             ^
make: *** [Release/obj.target/oracledb/src/njs/src/njsOracle.o] Error 1
make: Leaving directory `/home/dc-user/myproject/node_modules/oracledb/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_
modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_proces
s.js:200:12)
gyp ERR! System Linux 3.13.0-37-generic
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_mod
ules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/dc-user/myproject/node_modules/oracledb
gyp ERR! node -v v4.0.0
gyp ERR! node-gyp -v v3.0.1
gyp ERR! not ok
npm ERR! Linux 3.13.0-37-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "install" "oracledb"
npm ERR! node v4.0.0
npm ERR! npm  v2.14.2
npm ERR! code ELIFECYCLE
 
npm ERR! oracledb@1.1.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the oracledb@1.1.0 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the oracledb package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls oracledb
npm ERR! There is likely additional logging output above.
 
npm ERR! Please include the following file with any support request:
npm ERR!     /home/dc-user/myproject/npm-debug.log

**Comment by yunchiri** [comment]
I have same issue.

**Comment by cjbj** [comment]
For tracking: https://github.com/nodejs/node/issues/2798
It's early days for Node 4.

**Comment by yunchiri** [comment]
Sorry @cjbj  I didn't search before question .
Thanks @cjbj  goodday!

**Comment by derickbailey** [comment]
is there an ETA on this getting fixed? I'm trying to schedule server upgrades to Node v4, to coincide with project updates and am needing this project.

**Comment by justingregory** [comment]
@cjbj I'm working with @derickbailey at our customer's site today and they're asking us to open a SR through their official Oracle support channel (support.oracle.com) to determine when this issue will be resolved. 

This issue is critical to our customer and needs quick resolution.  Would opening a SR help expedite the resolution or do you expect it to be addressed soon through this channel?

Best regards,
Justin

**Comment by cjbj** [comment]
@justingregory @derickbailey node-oracledb has open source support, i.e. this is the right place to raise issues. I don't give ETAs because (i) there are too many variables that affect prediction (ii) there can be major consequences if we over-commit and under-deliver. 

I'd be happy to talk with your customer. Send me email (see my profile).

**Comment by doberkofler** [comment]
+1 (node 4.1.1 support would be very appreciated)

**Comment by Bigous** [comment]
I've made a pull request #227 upgrading Nan to 2.x.
It enables the usage of this driver with the latest node version (4.1.1) and continues to support all the previous one (from 0.10.28 to 4.1.1).
We need people testing it in more environments so @cjbj can merge soon.

**Comment by cjbj** [comment]
Thanks @Bigous.  As I mentioned in the PR, we will merge this when we get the go ahead - there are a few things we need to do first. Have patience.

**Comment by jeveloper** [comment]
+1 node 4.x support , thanks , you guys are doing great

**Comment by sagiegurari** [comment]
+1
I'll try to work with @Bigous version in the meantime.

**Comment by jeveloper** [comment]
Anyone tried LTS version of Node and this lib ? 4.2.X ?

**Comment by derickbailey** [comment]
@jeveloper my very simple scenarios are working well enough in Node 4.2.1 w/ @Bigous patch. admittedly, though, i have some *very* simple scenarios for my current use... open a connection, run a stored proc, close the connection. :)

**Comment by jeveloper** [comment]
Thanks @derickbailey , i'd prefer to avoid using a patch, i suppose we're looking at 1.4 of this lib right after Oracle conference , nov 1st ? :)))
:+1:

**Comment by pravincar** [comment]
+1 @cjbj Appreciate your concern on ETAs but it would be great if I could have a rough idea of the plan since I'll plan my development accordingly. Promise to not hold you accountable!

**Comment by cjbj** [comment]
@pravincar see https://github.com/oracle/node-oracledb/pull/227#issuecomment-151371363

**Comment by cj10** [comment]
+1
I need up-to-date versions of Node in order to use the newest es6 features in V8.
I am interested in using asynchronous generator functions to avoid callback hell in association with  oracledb.

**Comment by mrapczynski** [comment]
@cj10 I have used both generators and promises with oracledb to avoid callbacks, and I found promises to be the best solution - particularly if you use Bluebird, then you use their fancy using()/dispose() mechanism.

Initially, generators seemed like a great idea for async, but now that I've done it, using them to avoid callbacks feels like more of a hack. I look forward to the ES7 proposal for async/await to settle the matter for good.

**Comment by cjbj** [comment]
For people not following the PR: the current state of Node 4+ support is it is blocked by a NAN issue https://github.com/oracle/node-oracledb/pull/227#issuecomment-154362963

**Comment by bjouhier** [comment]
@cj10 @mrapczynski 
FWIW we are using streamline.js with oracledb, to avoid callback hell. 

We have been using it  for almost 5 years now, to develop a large node.js application without callbacks. It gives us the choice to generate code for raw callbacks, generators or fibers. Our experience is that the fibers option delivers best performance today because it implements deep continuations and it eliminates all the callback/generator/promise overhead in intermediate calls (async functions that just calls other async functions that we wrote - they dominate in our code). 

Disclamer: I wrote streamline.js so I'm obviously very biased.

**Comment by Raydekk** [comment]
+1
I can confirm the same bug for node v5.0.0

**Comment by cjbj** [comment]
@bjouhier are you interested in writing an article about using streamline.js & node-oracledb for OTN?

**Comment by bjouhier** [comment]
@cjbj That's a cool idea. I need a bit of validation first and I'll follow up with you by email (this is getting off-topic anyway).

**Comment by cjbj** [comment]
@bjouhier sounds good.

For everyone else, don't forget to +1 the NAN issue.

**Comment by bjouhier** [comment]
A big +1 to this issue BTW.

**Comment by cjbj** [comment]
@bjouhier I mean the NAN-NAN issue! https://github.com/nodejs/nan/issues/511

**Comment by bjouhier** [comment]
Got misled by the +1s above. Done!

**Comment by cjbj** [comment]
Status on this is that we have been stabilizing the functional test suite, waiting for stress tests to finish, evaluating whether to delay the release and squash compilation warnings seen in some environments, etc etc etc.  Basically, just making sure everything is fine and dandy.

**Comment by cjbj** [comment]
Done. Yay!  Thanks @Bigous.

**Comment by derickbailey** [comment]
+1000 huge thanks to @Bigous, @cjbj and everyone else that contributed to this! :)

**Comment by justingregory** [comment]
*high five*  Thanks again for the support!

**Comment by Bigous** [comment]
Tks guys, and @cjbj you gave a much better solution to the Reset() problem!

**Comment by cjbj** [comment]
@Bigous that would have been done by @pvenkatraman and co.
